src/main/tut/3.7-comonads.md

```diff
@@ -0,0 +1,453 @@
+  stm match {
+    case Stream(a, as) =>
+      import numeric._
+      if (n <= 0) zero else a() + sumS(n - 1)(as())
+  }
+```
+................
+```Haskell
+average :: Fractional a => Int -> Stream a -> a
+average n stm = (sumS n stm) / (fromIntegral n)
+```
+```tut:silent
+def average[A](n: Int)
+    (implicit fractional: Fractional[A])
+    : Stream[A] => A = stm => {
+  import fractional._
+  sumS(n)(stm) / fromInt(n)
+}
+```
+................
+```Haskell
+movingAvg :: Fractional a => Int -> Stream a -> Stream a
+movingAvg n = extend (average n)
+```
+```tut:silent
+def movingAvg[A](n: Int)(stm: Stream[A])
+    (implicit fractional: Fractional[A])
+    : Stream[A] =
+  streamComonad.
+    extend(average(n)(fractional))(stm)
+```
+................
+```Haskell
+class Comonoid m where
+  split   :: m -> (m, m)
+  destroy :: m -> ()
+```
+```tut:silent
+trait Comonoid[M] {
+  def split(x: M): (M, M)
+
+  def destroy(x: M): Unit
+}
+```
+................
+```Haskell
+destroy _ = ()
+```
+```scala
+def destroy(x: M): Unit = ()
+```
+................
+```Haskell
+split x = (f x, g x)
+```
+```scala
+def split(x: M): (M, M) = (f(x), g(x))
+```
+................
+```Haskell
+lambda . bimap destroy id . split = id
+rho . bimap id destroy . split = id
+```
+```scala
+(lambda compose bimap(destroy)(identity[M]))
+  .compose(split) == identity[M]
+
+(rho compose bimap(identity[M])(destroy))
+  .compose(split) == identity[M]
+```
+................
+```Haskell
+lambda (bimap destroy id (split x))
+= lambda (bimap destroy id (f x, g x))
+= lambda ((), g x)
+= g x
+```
+```scala
+lambda(bimap(destroy)(identity[M])(split(x))) ==
+  lambda(bimap(destroy)(identity[M])(
+    (f(x), g(x)))) ==
+  lambda(((), g(x))) ==
+  g(x)
+```
+................
+```Haskell
+split x = (x, x)
+```
+```scala
+def split(x: M): (M, M) = (x, x)
+```
+................
+```Haskell
+data Store s a = Store (s -> a) s
+```
+```tut:silent
+case class Store[S, A](run: S => A, s: S)
+```
+................
+```Haskell
+counit (Prod (Reader f, s)) = f s
+```
+```tut:silent
+def counit[S, A](a: Prod[S, Reader[S, A]])
+    : A = a match {
+  case Prod((Reader(f), s)) => f(s)
+}
+```
+................
+```Haskell
+extract (Store f s) = f s
+```
+```scala
+def extract[A](wa: Store[S, A])
+    : A = wa match {
+  case Store(f, s) => f(s)
+}
+```
+................
+```Haskell
+unit a = Reader (\s -> Prod (a, s))
+```
+```tut:silent
+def unit[S, A](a: A)
+    : Reader[S, Prod[S, A]] =
+  Reader(s => Prod((a, s)))
+```
+................
+```Haskell
+Store f :: s -> Store f s
+```
+```tut:silent
+object Store {
+  def apply[S, A](run: S => A)
+      : S => Store[S, A] =
+    s => new Store(run, s)
+}
+```
+```tut:invisible
+// for tut: need both case class and object
+object trick {
+  object Store {
+    def apply[S, A](run: S => A)
+        : S => Store[S, A] =
+      s => new Store(run, s)
+  }
+  case class Store[S, A](run: S => A, s: S)
+}
+import trick._
```

Comment:
The only idea I had for having both case class and companion object at the same time to make further snippets compilable.